src/py/mat3ra/made/tools/calculate/calculators.py

```diff
@@ -120,3 +123,112 @@ def get_energy(
         substrate_coordinates_values = np.array(substrate_surface_atom_coordinates.values)
 
         return interaction_function(film_coordinates_values, substrate_coordinates_values)
+
+
+class FilmSubstrateDistanceASECalculator(ASECalculator):
+    """
+    ASE calculator that calculates the interaction energy between a film and substrate in an interface material.
+
+    Args:
+        shadowing_radius (float): The radius for atom to shadow underlying from being considered surface, in Angstroms.
+        force_constant (float): The force constant for the finite difference approximation of the forces.
+        is_substrate_fixed (bool): Whether to fix the substrate atoms.
+        is_z_axis_fixed (bool): Whether to fix atoms movement in the z direction.
+        symprec (float): The symmetry precision for the ASE calculator. This parameter determines the tolerance
+                         for symmetry operations, affecting the identification of equivalent atoms and the overall
+                         symmetry of the system. For more details, refer to the ASE documentation:
+                         https://wiki.fysik.dtu.dk/ase/ase/constraints.html#ase.constraints.FixSymmetry
+
+    Example usage:
+    ```python
+    from ase.optimize import BFGS
+    atoms = to_ase(material)
+    calc = SurfaceDistanceCalculator(shadowing_radius=2.5)
+
+    atoms.calc = calc
+    opt = BFGS(atoms)
+    opt.run(fmax=0.05)
+    ```
+    Args:
+        shadowing_radius (float): Radius for atoms shadowing underlying from being treated as a surface, in Angstroms.
+        force_constant (float): The force constant for the finite difference approximation of the
+    Note:
+        Built following https://wiki.fysik.dtu.dk/ase/development/calculators.html
+
+        The calculate method is responsible for computing the energy and forces (if requested).
+        Forces are estimated using a finite difference method, which is a simple approximation
+        and might not be the most accurate or efficient for all cases.
+    """
+
+    implemented_properties = ["energy", "forces"]
+
+    def __init__(
+        self,
+        shadowing_radius: float = 2.5,
+        is_substrate_fixed: bool = True,
+        is_z_axis_fixed: bool = True,
+        symprec: float = 0.01,
+        material_calculator: Union[MaterialCalculator, InterfaceMaterialCalculator] = InterfaceMaterialCalculator(),
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.shadowing_radius = shadowing_radius
+        self.fix_substrate = is_substrate_fixed
+        self.fix_z = is_z_axis_fixed
+        self.symprec = symprec
+        self.material_calculator = material_calculator
+
+    def _add_constraints(self, atoms: ASEAtoms) -> ASEAtoms:
+        constraints: List[Union[ASEFixAtoms, ASEFixedPlane, RigidFilmXYConstraint]] = []
+        if self.fix_substrate:
+            substrate_indices = [i for i, tag in enumerate(atoms.get_tags()) if tag == 0]
```

Comment:
We should abstract this into `get_interface_part_indices`

Reply:
Ok, but locally. Since it's ASE atoms, not Material